src/openrct2/world/Wall.cpp

```diff
@@ -701,21 +701,45 @@ extern "C"
         wallElement->properties.wall.animation |= (frameNum & 0xF) << 3;
     }
 
-    uint8 wall_element_get_secondary_colour(rct_tile_element * wallElement)
+    colour_t wall_get_primary_colour(const rct_tile_element * tileElement)
     {
-        uint8 secondaryColour = (wallElement->properties.wall.colour_1 & 0xE0) >> 5;
+        return tileElement->properties.wall.colour_1 & TILE_ELEMENT_COLOUR_MASK;
+    }
+
+    colour_t wall_get_secondary_colour(rct_tile_element * wallElement)
+    {
+        uint8 secondaryColour = (wallElement->properties.wall.colour_1 &~ TILE_ELEMENT_COLOUR_MASK) >> 5;
         secondaryColour |= (wallElement->flags & 0x60) >> 2;
         return secondaryColour;
     }
 
-    void wall_element_set_secondary_colour(rct_tile_element * wallElement, uint8 secondaryColour)
+    colour_t wall_get_tertiary_colour(const rct_tile_element * tileElement)
+    {
+        return tileElement->properties.wall.colour_3 & TILE_ELEMENT_COLOUR_MASK;
+    }
+
+    void wall_set_primary_colour(rct_tile_element * tileElement, colour_t colour)
     {
-        wallElement->properties.wall.colour_1 &= 0x1F;
+        assert(colour <= 31);
+        tileElement->properties.wall.colour_1 &= ~TILE_ELEMENT_COLOUR_MASK;
+        tileElement->properties.wall.colour_1 |= colour;
+    }
+
+    void wall_set_secondary_colour(rct_tile_element * wallElement, colour_t secondaryColour)
+    {
+        wallElement->properties.wall.colour_1 &= TILE_ELEMENT_COLOUR_MASK;
```

Comment:
~TILE_ELEMENT_COLOUR_MASK

Reply:
No, this is correct.. It should keep colour 1, and delete the colour 2 info.